src/org/sipdroid/sipua/PublishAgent.java

```diff
@@ -0,0 +1,213 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this source code; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ * Author(s):
+ * Pranav Jain (contact@pranavjain.me)
+ */
+
+package org.sipdroid.sipua;
+
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.preference.PreferenceManager;
+
+import org.zoolu.sip.address.NameAddress;
+import org.zoolu.sip.authentication.DigestAuthentication;
+import org.zoolu.sip.header.AuthorizationHeader;
+import org.zoolu.sip.header.ProxyAuthenticateHeader;
+import org.zoolu.sip.header.ProxyAuthorizationHeader;
+import org.zoolu.sip.header.StatusLine;
+import org.zoolu.sip.header.ViaHeader;
+import org.zoolu.sip.header.WwwAuthenticateHeader;
+import org.zoolu.sip.message.Message;
+import org.zoolu.sip.message.MessageFactory;
+import org.zoolu.sip.provider.SipProvider;
+import org.zoolu.sip.provider.SipStack;
+import org.zoolu.sip.transaction.TransactionClient;
+import org.zoolu.sip.transaction.TransactionClientListener;
+import org.zoolu.tools.BasicTimerFactory;
+import org.zoolu.tools.LogLevel;
+
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.text.MessageFormat;
+import java.util.logging.Logger;
+
+
+public class PublishAgent implements TransactionClientListener {
+
+	/**
+	 * UserProfile
+	 */
+	protected UserAgentProfile user_profile;
+	/**
+	 * User name.
+	 */
+	String username;
+	/**
+	 * User realm.
+	 */
+	String realm;
+	/**
+	 * Nonce for the next authentication.
+	 */
+	String next_nonce;
+	/**
+	 * User's passwd.
+	 */
+	String passwd;
+
+	private Logger logger = Logger.getLogger(getClass().getCanonicalName());
+	/**
+	 * SipProvider
+	 */
+	protected SipProvider sip_provider;
+	Context context;
+
+	public PublishAgent(SipProvider sip_provider, UserAgentProfile user_profile, String username, String realm, String passwd, Context context) {
+		this.sip_provider = sip_provider;
+		this.user_profile = user_profile;
+		this.username = username;
+		this.realm = realm;
+		this.passwd = passwd;
+		this.next_nonce = null;
+		this.context = context;
+
+	}
+
+	public void publish() {
+		this.publish(BasicStatus.OPEN, SipStack.default_expires, "");
+	}
+
+	public void publish(BasicStatus status, int expireTime, String note) {
+		SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
+		Boolean publish_enable_status = prefs.getBoolean("publish_enable", true);
+		if (publish_enable_status == true) {
+			MessageDigest md = null;
+			String tupleId;
+			try {
+				md = MessageDigest.getInstance("MD5");
+				tupleId =  md.digest(user_profile.username.getBytes()).toString();

```

Comment:
or even better, just implement the code to convert to a hex string.  Calling ```toString()``` here is not actually returning the contents of the ```byte[]```